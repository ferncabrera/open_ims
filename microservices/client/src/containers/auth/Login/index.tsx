import React, { SyntheticEvent, useState, useEffect } from 'react'
import { Link } from 'react-router-dom';
import { Form, Row, Button, Spinner } from 'react-bootstrap'
import { Password } from '../../../components/forms/Password';
import { CenteredModal } from '../../../components/modals/CenteredModal';
import { MdOutlineEmail } from 'react-icons/md';
import { fieldValidation } from '../../../utilities/validation';
import { IValidate } from '../../../utilities/types/validationTypes';
import { sendPostRequest, getJSONResponse } from '../../../utilities/apiHelpers';
import { useNavigate } from 'react-router-dom';
import { bannerState } from '../../../atoms/atoms';
import { useAtom } from 'jotai';
import { useResetAtom } from 'jotai/utils';
import _ from 'lodash';

import styles from './index.module.scss';

const initialErrorFieldState = {
    email: null,
    password: null,
}

const formData = {};
let trackErrorList : Array<boolean> = [];


export const Login = () => { // test comment


    const keyPaths = {
        attributes: 'attributes',
        email: 'attributes.email',
        password: 'attributes.password',
    }

    const [showResetPassword, setShowResetPassword] = useState(false);
    const [emailSent, setEmailSent] = useState(false);
    const [error, setError] = useState<IErrorFields>(initialErrorFieldState);
    const [resetEmail, setEmail] = useState('');
    const [isSigningIn, setIsSigningIn] = useState(false);
    const [bannerText, setBannerTextState] = useAtom(bannerState);
    const resetBannerState = useResetAtom(bannerState);
    const navigate = useNavigate();

    useEffect(() => {
        _.set(formData, keyPaths.email, "");
        _.set(formData, keyPaths.password, "");
    }, []);

    useEffect(() => {
        if (!showResetPassword) {
            // Reset the email state when the form is not shown
            setEmail('');
        }
    }, [showResetPassword]);

    const modalFooter = emailSent ? <></> : <a
        onClick={() => {
            setShowResetPassword(false);
            setEmailSent(false);
        }}
    >
        Cancel
    </a>


    const handleForgotPass = async (e: SyntheticEvent) => {
        setEmailSent(true);
        e.preventDefault(); // Prevents the default form submission behavior
        // Send post request here
        await sendPostRequest({ endpoint: "/api/server/forgot_pass", data: { email: resetEmail } });

    }

    const resetPasswordContent = () => {
        return (
            <div>
                <h6 className='mb-3'>Enter the email associated with you CCG inventory account here:</h6>
                <Form noValidate onSubmit={handleForgotPass}>
                    <Form.Control type="email" placeholder='' value={resetEmail} onChange={(e) => { setEmail(e.target.value) }} />
                    <Button
                        type='submit'
                        id="send-email"
                        className='mt-2'>
                        Send email
                    </Button>
                </Form>
                <p className='mt-5 text-center'>If we find your email in our records, we'll send you an email with next steps.</p>
            </div>
        )
    };

    const resetPasswordSentContent = () => {
        return (
            <div>
                <h2 className='mb-5 pb-5 d-flex justify-content-center'>Email Sent!</h2>
                <div className='mt-5'>
                    <span className='d-flex justify-content-center'>
                        <MdOutlineEmail className="font-24" />
                    </span>
                    <p className='text-center'>Keep an eye out for an incoming email with your next steps.</p>
                </div>
            </div>
        )
    };

    const validate = (data: IValidate) => {
        const valid = fieldValidation(data)
        const name = valid.fieldName;
        if (!valid.isValid) {
            setError((prevError) => ({ ...prevError, [name]: { valid: valid.isValid, message: valid.message } }));
        } else {
            setError((prevError) => ({ ...prevError, [name]: null }))
        }
        return valid.isValid;
    }

    const handleSignIn = async (e: SyntheticEvent) => {
        e.preventDefault();
        const loginData = _.get(formData, keyPaths.attributes);
        _.forIn(loginData, (value, key: string) => {
            const data = { value, name: key, required: true }
            const isValid = validate(data);
            trackErrorList.push(isValid)
        });
        if (_.some(trackErrorList, (validEntity) => validEntity === false)) {
            trackErrorList = [];
            return
        }
        // Send post request here
        try {

            setIsSigningIn(true);
            const response: IResponse = await sendPostRequest({ endpoint: '/api/server/login', data: loginData });
            if (_.get(response, 'token', '')) {
                navigate('/ccg/dashboard');
                resetBannerState();
            } else {
                setBannerTextState({variant: 'danger', message: response?.message ? response.message : global.__APP_DEFAULT_ERROR_MSG__});
            }
            setIsSigningIn(false);
        } catch (e) {
            //need our global state banner here to handle our errors
            setBannerTextState({variant: 'danger', message: (e?.toString()?e.toString(): global.__APP_DEFAULT_ERROR_MSG__)});
        }
    }

    return (
        <>
            <div className='mt-5 mb-5'>
                <h1>Welcome back</h1>
                <div className={`d-flex justify-content-start ${styles.flex_wrapper}`}>
                    <h6 className='d-flex align-items-center'>New to CCG Inventory?</h6>
                    <Link className={`ps-1 ${styles.link_header}`} to={'/register'}> Create an account</Link>
                </div>
            </div>
            <Form noValidate onSubmit={handleSignIn}>
                <Form.Group className='mb-3'>
                    <Form.Label htmlFor='email'>Email:</Form.Label>
                    <Form.Control
                        name="email"
                        id='email'
                        type="email"
                        placeholder=''
                        isInvalid={error.email?.valid === false}
                        required
                        onChange={(e) => { _.set(formData, keyPaths.email, e.target.value) }}
                        onBlur={(data) => {
                            const target = data.target as HTMLTextAreaElement;
                            const { value, name, required } = target;
                            validate({ value, name, required })
                        }}
                    />
                    {error.email?.valid === false &&
                        <Form.Control.Feedback data-testid="email-error" type='invalid'>{error.email.message}</Form.Control.Feedback>
                    }
                </Form.Group>
                <Password
                    isInvalid={error.password?.valid === false}
                    onChange={(e) => { _.set(formData, keyPaths.password, e.target.value) }}
                    onBlur={(data: SyntheticEvent) => {
                        const target = data.target as HTMLTextAreaElement;
                        const { value, name, required } = target;
                        validate({ value, name, required });
                    }}
                    required
                    errorMessage={_.get(error, "password.message", '')}
                />

                <Row>
                    <div className={'d-flex justify-content-between' + ` ${styles.flex_wrapper}`}>
                        <Form.Check label='Remember me' />
                        <a className='no-underline' onClick={() => setShowResetPassword(true)}>forgot password</a>
                    </div>
                </Row>
                <div className='mt-5'>
                    <Button
                        type='submit'
                        disabled={isSigningIn}
                    >
                        {isSigningIn ? <Spinner animation='border'/> : "Sign in"}
                    </Button>
                </div>
            </Form>
            {/* Display Modal here */}
            <CenteredModal
                heading={<div><h3>Forgot your Password?</h3><h3>No Worries.</h3></div>}
                show={showResetPassword}
                footer={modalFooter}
                topCloseFunction={() => {
                    setShowResetPassword(false)
                    setEmailSent(false)
                }}
            >
                {!emailSent ? resetPasswordContent() : resetPasswordSentContent()}
            </CenteredModal>
        </>
    )
}
