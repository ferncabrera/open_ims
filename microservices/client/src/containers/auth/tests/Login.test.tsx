import React from "react";
import { expect, describe, it } from 'vitest';
import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { MemoryRouter } from "react-router";
//?  Pretty sure we don't even need a provider but just shimming one in here since we had one for the Recoil tests. Not sure if by convention or testing 
//?  environment necessity...
import { Provider } from "jotai";
import { Login } from "../Login";

function tree() {
    return render(<MemoryRouter><Provider><Login /></Provider></MemoryRouter>)
}



describe("Login Component", () => {

    // This first test case takes a snapshot. Compares and makes sure no previous styles or html changes were made from its last snapshot.
    it("Login renders correctly", () => {
        const tree = renderer.create(<MemoryRouter><Provider><Login /></Provider></MemoryRouter>).toJSON();
        expect(tree).toMatchSnapshot();

    });

    //Unit Test Cases for components

    it("Test reset password modal functionality", () => {
        // Here is an example of a DOM test. Ensuring that certain html and operations on what the user sees is functional and is expected.
        const { getByText, queryByText } = tree();

        expect(queryByText("Send email")).not.toBeInTheDocument(); // modal should not be open!
        fireEvent.click(getByText("forgot password")); // we click forgot password
        expect(getByText("Send email")).toBeInTheDocument(); // we check to see if the modal displays
        expect(queryByText("Email Sent!")).not.toBeInTheDocument(); // This should not show until we sent the email
        fireEvent.click(getByText("Send email")); // send email
        expect(queryByText("Send email")).not.toBeInTheDocument(); // modal screen should have changed!
        expect(getByText("Email Sent!")).toBeInTheDocument(); // Our final modal screen

    });

    it("Test validation function", () => {
        // Here we test the functionality of validations as a user fills out the form
        const { getByText, getByLabelText } = tree();
        const emailInput = getByLabelText("Email:");
        const passwordInput = getByLabelText("Password:")

        fireEvent.change(emailInput, { target: { value: "" } });
        fireEvent.blur(emailInput);


        expect(getByText("Required")).toBeInTheDocument();

        fireEvent.change(emailInput, { target: { value: "invalid-email" } });
        fireEvent.blur(emailInput);

        expect(getByText("Please enter a valid email")).toBeInTheDocument();

        fireEvent.change(passwordInput, { target: { value: "" } });
        fireEvent.blur(passwordInput);


        expect(getByText("Required")).toBeInTheDocument();

    });

    it("Test handle sign in", async () => {
        const { getByLabelText, getByText, queryByTestId } = tree();
        const emailInput = getByLabelText("Email:");
        const passwordInput = getByLabelText("Password:");
        const signInButton = getByText("Sign in");

        const testInputCases = [
            { email: "", password: "" },
            { email: "", password: "hello123" },
            { email: "oliver", password: "hello123" },
            { email: "oliver@gamil.com", password: "" },
            { email: "oliver@gmail.com", password: "hello123" }
        ];

        const testCases = [
            { emailV: "Required", passwordV: "Required" },
            { emailV: "Required", passwordV: "" },
            { emailV: "Please enter a valid email", passwordV: "" },
            { emailV: "", passwordV: "Required" },
            { emailV: "", passwordV: "" }
        ];

        for (let i = 0; i < testCases.length; i += 1) {
            fireEvent.change(emailInput, { target: { value: testInputCases[i].email } });
            fireEvent.change(passwordInput, { target: { value: testInputCases[i].password } });
            fireEvent.click(signInButton);

            if (testCases[i].emailV) {
                expect(queryByTestId("email-error")).toBeInTheDocument();
            } else {
                expect(queryByTestId("email-error")).not.toBeInTheDocument();
            }

            if (testCases[i].passwordV) {
                expect(queryByTestId("password-error")).toBeInTheDocument();
            } else {
                expect(queryByTestId("password-error")).not.toBeInTheDocument();
            }
        }

        //check if a cookie is created from our mock apiCall
        await new Promise((resolve) => setTimeout(resolve, 100)); // Allow asynchronous actions to create cookie

        expect(document.cookie).toContain('is-authenticated=true'); // Check if the cookie is present in document.cookie


    });


})