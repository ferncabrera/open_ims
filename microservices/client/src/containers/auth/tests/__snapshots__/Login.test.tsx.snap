// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login Component > Login renders correctly 1`] = `
[
  <div
    className="mt-5 mb-5"
  >
    <h1>
      Welcome back
    </h1>
    <div
      className="d-flex justify-content-start _flex_wrapper_66e569"
    >
      <h6
        className="d-flex align-items-center"
      >
        New to CCG Inventory?
      </h6>
      <a
        className="ps-1 _link_header_66e569"
        href="/register"
        onClick={[Function]}
      >
         Create an account
      </a>
    </div>
  </div>,
  <form
    className=""
    noValidate={true}
    onSubmit={[Function]}
  >
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="email"
      >
        Email:
      </label>
      <input
        className="form-control"
        id="email"
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder=""
        required={true}
        type="email"
      />
    </div>
    <label
      className="form-label"
      htmlFor="password"
    >
      Password:
    </label>
    <div
      style={
        {
          "marginBottom": "20px",
        }
      }
    >
      <div
        className="input-group"
        style={
          {
            "display": "block",
            "marginBottom": "0px",
          }
        }
      >
        <input
          className="_password_8ab8af undefined  form-control"
          id="password"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          type="password"
        />
        <span
          className="_password_8ab8af"
          onClick={[Function]}
        >
          <svg
            className="font-24"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M12 6a9.77 9.77 0 018.82 5.5C19.17 14.87 15.79 17 12 17s-7.17-2.13-8.82-5.5A9.77 9.77 0 0112 6m0-2C7 4 2.73 7.11 1 11.5 2.73 15.89 7 19 12 19s9.27-3.11 11-7.5C21.27 7.11 17 4 12 4zm0 5a2.5 2.5 0 010 5 2.5 2.5 0 010-5m0-2c-2.48 0-4.5 2.02-4.5 4.5S9.52 16 12 16s4.5-2.02 4.5-4.5S14.48 7 12 7z"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="d-flex justify-content-between _flex_wrapper_66e569"
      >
        <div
          className="form-check"
        >
          <input
            className="form-check-input"
            disabled={false}
            type="checkbox"
          />
          <label
            className="form-check-label"
            title=""
          >
            Remember me
          </label>
        </div>
        <a
          className="no-underline"
          onClick={[Function]}
        >
          forgot password
        </a>
      </div>
    </div>
    <div
      className="mt-5"
    >
      <button
        className="btn btn-primary"
        disabled={false}
        type="submit"
      >
        Sign in
      </button>
    </div>
  </form>,
]
`;
